### PR TITLE
Implement the DiceKeys API

### PR DESCRIPTION
**Pull request by UppaJung**
Make api testable
Write first test for API

### CODE REVIEW

File: DiceKeys/State/DiceKeyMemoryStore.swift [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+//
+//  Created by Stuart Schechter on 2021/02/08.
+//
+
+import Foundation
+
+final class DiceKeyMemoryStore: ObservableObjectUpdatingOnAllChangesToUserDefaults {
+    static private(set) var singleton = DiceKeyMemoryStore()
+    
+    init(_ diceKeyLoaded: DiceKey? = nil) {
```
Comment (Abhityagi16) [inline DiceKeys/State/DiceKeyMemoryStore.swift:13]: This constructor defeats the purpose of singleton. Either remove singleton or remove this construction. We can move this to a variable `var diceKeyLoaded: DiceKey?`

Comment (UppaJung) [reply DiceKeys/State/DiceKeyMemoryStore.swift:13]: Except that for previewing/testing we need to have different values within the same process, so while the app uses a singleton, the previews do not.

Is there a standard way to handle it?  I'll put a comment in for now.

**Comment by UppaJung** [comment]
@Abhityagi16 , if you don't see any further changes after reviewing, please approve the merge.

image

File: DiceKeys/Core/Api/AuthenticationTokens.swift [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+//
+//  Created by Stuart Schechter on 2021/02/04.
+//
+
+import Foundation
+
+class AuthenticationTokens {
+    static var tokenToUrl: [String: String] = [:]
+
+    static func setToken(authToken: String, replyToUrl: String) {
```
Comment (Abhityagi16) [inline DiceKeys/Core/Api/AuthenticationTokens.swift:13]: Where are we using this function to set the token? Is it not in this PR?

**Comment by UppaJung** [comment]
Not yet.

File: DiceKeys/Core/Api/BackgroundCalculationOfApiRequestResult.swift [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        let cacheKey = SHA256.hash(data: cacheKeyPreimage.data(using: .utf8)!).description
+        var result: Result<SuccessResponse,Error>?
+        do {
+            try result = .success(request.execute(seedString: seedString))
+        } catch {
+            result = .failure(error)
+        }
+        
+        let resultFuture = Future<SuccessResponse, Error>{ promise in promise(result!) }
+        let r = BackgroundCalculationOfApiRequestResult.init(resultFuture)
```
Comment (Abhityagi16) [inline DiceKeys/Core/Api/BackgroundCalculationOfApiRequestResult.swift:57]: nit: `BackgroundCalculationOfApiRequestResult.init(resultFuture)` can be replaced with `BackgroundCalculationOfApiRequestResult(resultFuture)`

File: DiceKeys/Core/Api/Unmarshalling.swift [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        if let queryItems = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems {
+            for queryItem in queryItems {
+                queryDictionary[queryItem.name] = queryItem.value
+            }
+        }
+        self.parameters = queryDictionary
+    }
+
+    func optionalField(name fieldName: String) -> String? {
+        return parameters[fieldName] ?? nil
```
Comment (Abhityagi16) [inline DiceKeys/Core/Api/Unmarshalling.swift:29]: ?? doesn't serve any purpose. If `parameters[fieldName]` returns `nil` then `return nil` doesn't seem right

File: DiceKeys/Core/Api/Unmarshalling.swift [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    }
+    func requiredField(name fieldName: String) throws -> String {
+        let value = parameters[fieldName] ?? nil
+        guard let nonNilValue = value else {
```
Comment (Abhityagi16) [inline DiceKeys/Core/Api/Unmarshalling.swift:33]: Replace with
```
guard let nonNilValue = parameters[fieldName] else {
...
}
```

File: DiceKeys/Core/Api/Unmarshalling.swift [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+protocol ApiRequestParameterUnmarshaller {
+    func optionalField(name fieldName: String) -> String?
+    func requiredField(name fieldName: String) throws -> String
+}
+
+class UrlParameterUnmarshaller: ApiRequestParameterUnmarshaller {
+    private let parameters: [String: String?]
+
+    init(url: URL) {
+        var queryDictionary = [String: String?]()
```
Comment (Abhityagi16) [inline DiceKeys/Core/Api/Unmarshalling.swift:19]: I would suggest making the dictionary value non-optional [String: String]. Maintaining nil values in memory is just overhead. If a key is not found in dictionary then it returns nil anyway

File: DiceKeys/Core/Api/AuthenticationTokens.swift (hunk 2 above)
Comment (UppaJung) [reply DiceKeys/Core/Api/AuthenticationTokens.swift:13]: Not in this PR.  Will use later.

File: DiceKeys/Core/Api/BackgroundCalculationOfApiRequestResult.swift (hunk 3 above)
Comment (UppaJung) [reply DiceKeys/Core/Api/BackgroundCalculationOfApiRequestResult.swift:57]: Good point.  Fixing.

File: DiceKeys/Core/Api/Unmarshalling.swift (hunk 4 above)
Comment (UppaJung) [reply DiceKeys/Core/Api/Unmarshalling.swift:29]: Fixing

File: DiceKeys/Core/Api/Unmarshalling.swift (hunk 5 above)
Comment (UppaJung) [reply DiceKeys/Core/Api/Unmarshalling.swift:33]: Fixing

File: DiceKeys/Core/Api/Unmarshalling.swift (hunk 6 above)
Comment (UppaJung) [reply DiceKeys/Core/Api/Unmarshalling.swift:19]: Fixing

File: DiceKeys/Core/Api/Unmarshalling.swift [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+import Foundation
+
+protocol ApiRequestParameterUnmarshaller {
+    func optionalField(name fieldName: String) -> String?
+    func requiredField(name fieldName: String) throws -> String
+}
+
+class UrlParameterUnmarshaller: ApiRequestParameterUnmarshaller {
+    private let parameters: [String: String?]
```
Comment (Abhityagi16) [inline DiceKeys/Core/Api/Unmarshalling.swift:16]: Replace this as well with `[String: String]`

**Pull request merged**